### ISSUE TITLE
Using audio DTMF increases the delay of the outgoing audio stream...

### ISSUE THREAD

**Issue by osenenko-e** [baresip/baresip#1703]
I use "dtmfmode=auto" (DTMFMODE_AUTO)
when I call function "int call_send_digit(struct call *call, char key)"
the outgoing audio stream delay is increased on DTMF digit send time...
It seems that the DTMF audio data is just inserted into the outgoing stream, although we need to remove the original source data on the duration of the DTMF digit
Tested in baresip console application compiled for Windows...
Any idea?

**Comment by alfredh** [comment]
https://github.com/baresip/baresip/wiki/Guidelines-for-reporting-Issues

please provide the necessary details like config and logs

**Comment by osenenko-e** [comment]
You can download the compiled version from here:
ftp://www.mnppsaturn.ru/osenenko/baresipbcb.zip
The config is in "config" subdirrectory
Full source is located here:
ftp://www.mnppsaturn.ru/osenenko/BareSipBcb_fullsrc.zip
Both zips have password "baresip"
This is the console application (baresip 1.1) compiled in BCB2007
How to see the problem:
1. Start conversation
2. Press keys 0-9 to send audio DTMF (dtmfmode=auto is set in accounts)
Each pressing increase the outgoung audio stream delay about 400 ms (I guess it depends on TXC_DIGIT_MIN)

**Comment by osenenko-e** [comment]
accounts:
<sip:212@sip.mnppsaturn.ru>;auth_pass=******;dtmfmode=auto
config:
# Core
poll_method		select		# poll, select

# SIP
sip_listen		0.0.0.0:5061
sip_transports		udp
sip_tos			160

# Call
call_local_timeout	120
call_max_calls		4
call_hold_other_calls	yes

# Audio
audio_player		winwave,"Динамики (E-MU 0202 | USB)     "
audio_source		winwave,nil
audio_alert		winwave,nil
audio_level		no
ausrc_format		s16		# s16, float, ..
auplay_format		s16		# s16, float, ..
auenc_format		s16		# s16, float, ..
audec_format		s16		# s16, float, ..
audio_buffer		20-160		# ms
audio_telev_pt		101		# payload type for telephone-event

# Video
video_size		640x480
video_bitrate		1000000
video_fps		30.00
video_fullscreen	no
videnc_format		yuv420p

# AVT - Audio/Video Transport
rtp_tos			184
rtp_video_tos		136
rtcp_mux		no
jitter_buffer_type	fixed		# off, fixed, adaptive
jitter_buffer_delay	5-10		# frames
rtp_stats		no
#------------------------------------------------------------------------------
# Modules
# UI Modules
module			wincons.dll
module			g711.dll
module			winwave.dll
module			dshow.dll
module			uuid.dll
# Media NAT modules
module			stun.dll
module			turn.dll
module			ice.dll
#------------------------------------------------------------------------------
# Application Modules
module_app		account.dll
module_app		contact.dll
module_app		debug_cmd.dll
module_app		menu.dll
#------------------------------------------------------------------------------
# Module parameters
# UI Modules parameters
cons_listen		0.0.0.0:5555 # cons - Console UI UDP/TCP sockets
http_listen		0.0.0.0:8000 # httpd - HTTP Server
ctrl_tcp_listen		0.0.0.0:4444 # ctrl_tcp - TCP interface JSON
evdev_device		/dev/input/event0
# Opus codec parameters
opus_bitrate		28000 # 6000-510000

vumeter_stderr		yes

# Selfview
video_selfview		window # {window,pip}

**Comment by alfredh** [comment]
thanks ! I have some more questions for you:

- could you also include the log file and/or wireshark for the relevant session?

- which SIP client are you testing with ?

- would it be possible to estimate the amount of delay that each DTMF tone adds ?

**Comment by alfredh** [comment]
I have tested locally with 2 baresip's back-to-back. The latency is very low.
After sending hundreds of DTMF events, the latency is still low.
It does not seem to be possible to reproduce this here.

can you please test with a different SIP client and your SIP network ?

We should try to find out where the delay is coming from:

1. sender
2. server
3. receiver

**Comment by osenenko-e** [comment]
OK Great news!
I compiled the latest version of the console baresip application for Windows and tested it...
I also tested the p2p mode (baresip to baresip without a server)!
And this is what I noticed:
account option ";dtmfmode=auto;" for dtmf code does not work...
Baresip application does not transmit DTMF as a tone sound in the stream (the remote client does not hear it)
Can you check again whether the DTMF tone sounds are heard during the conversation!
Sorry for bother...

**Comment by osenenko-e** [comment]
And one more question...
I can't find where the DTMF audio is generated in the source code...
There is a function for generating DTMF audio digits - this is:
 int autone_dtmf(struct buf *mb, uint32_t state, int digit)
But it is not called anywhere
There are also sound files "/share/soundX.wav"
But I also can't find how they are used...
Thank you

**Comment by alfredh** [comment]
I have tested locally using 2 baresip's and dtmfmode=auto (rtpevent).

A:

```
audio: send DTMF digit: '1'    
audio: send DTMF digit end: '1'
audio: send DTMF digit: '2' (bit/s)    
audio: send DTMF digit end: '2't/s)    
audio: send DTMF digit: '3' (bit/s)    
audio: send DTMF digit end: '3't/s)    
audio: send DTMF digit: '4' (bit/s)    
audio: send DTMF digit end: '4't/s)    
audio: send DTMF digit: '5' (bit/s)    
audio: send DTMF digit end: '5't/s)    
```

B:
```
received in-band DTMF event: '1' (end=0)
received in-band DTMF event: '1' (end=1)
received in-band DTMF event: '2' (end=0)
received in-band DTMF event: '2' (end=1)
received in-band DTMF event: '3' (end=0)
received in-band DTMF event: '3' (end=1)
received in-band DTMF event: '4' (end=0)
received in-band DTMF event: '4' (end=1)
received in-band DTMF event: '5' (end=0)
received in-band DTMF event: '5' (end=1)
```

as you can see, it is working. In-band here means in-band in the RTP stream.
The received tones are not meant for human consumption, but for machine consumption.

**Comment by osenenko-e** [comment]
Ок I see the same:
received in-band DTMF event:  '8' (end=0)
received in-band DTMF event:  '8' (end=1)
"in-band in the RTP stream"  - the DTMF message in this mode is special digital telefone events...
But I want to send the DTMF as audio tone !
How to switch the console application to audio DTMF mode?

**Comment by mbattista** [comment]
> But I want to send the DTMF as audio tone !

If I look into the code correctly baresip only supports DTMF tones via `RTP` or `SIP info`. 
Real "in audio" DTFM tones are not supported, out of date and should not be used.

Why do you need "in audio" DTFM tones?
Just as acoustic feedback that it works or do you have an old machine which is controlled by those DTFM tones?

On your question why the `/share/soundX.wav` files exists. They are used by the gtk module as feedback since it does not have the visual feedback that the menu delivers.

**Comment by osenenko-e** [comment]
Thanks a lot for the explanation...
Yesterday when I tested the p2p baresip-baresip 
I realized that baresip does not transmit DTMF in the form of audio ...
And then everything became clear...
Earlier I tested "baresip<->Asterisk PBX<->PhonerLite"
and now it is clear that the problem is in the PhonerLite client
When PhonerLite receives DTMF in-band event start/end, it starts inserting audio of this DTMF code
into the audio stream, and the problem is that PhonerLite does not replace audio in the received stream, 
but simply inserts it, which leads to a delay in the main sound stream by the amount of the generated DTMF audio.
Once again, I apologize for the trouble!
Baresip is working correctly and it's the time to close this issue...
I need DTMF audio because there are still analog PBX.
But this is not a problem - I can always insert DTMF audio into the outgoing audio stream in "winwave"
Thanks again to everyone!
